File: src/tsuuid/dimensions.py
"""
tsuuid.dimensions — The 81-Dimensional Semantic Axis Definitions

Each dimension represents an independent axis of meaning.
Trit values: -1 (negative pole), 0 (neutral/absent), +1 (positive pole)

IMPORTANT: These axes are preliminary and require empirical validation.
This is where academic collaboration is most needed. The axes should
be validated against diverse domain schemas to ensure universal coverage.

Reference: TSUUID Framework paper, Section 3.2 (Hay, 2026)
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple
import numpy as np


@dataclass
class Axis:
    """A single semantic dimension."""
    id: int
    name: str
    negative: str   # meaning of trit = -1
    neutral: str    # meaning of trit = 0
    positive: str   # meaning of trit = +1
    layer: str      # trust hierarchy layer this belongs to
    
    def describe_trit(self, value: int) -> str:
        if value == -1:
            return f"{self.name}: {self.negative}"
        elif value == 0:
            return f"{self.name}: {self.neutral}"
        elif value == 1:
            return f"{self.name}: {self.positive}"
        raise ValueError(f"Invalid trit value: {value}")


# ============================================================
# PRELIMINARY AXIS DEFINITIONS
# Organized by trust hierarchy layer (trits 1-81)
# ============================================================

# Layer: Protocol / Standards (trits 1-10)
# These dimensions encode structural/protocol-level semantics
PROTOCOL_AXES = [
    Axis(1,  "temporality",      "past",         "atemporal",     "future",        "protocol"),
    Axis(2,  "modality",         "deletion",      "reference",     "creation",      "protocol"),
    Axis(3,  "certainty",        "denied",        "unknown",       "confirmed",     "protocol"),
    Axis(4,  "cardinality",      "singular",      "unspecified",   "plural",        "protocol"),
    Axis(5,  "obligation",       "forbidden",     "optional",      "required",      "protocol"),
    Axis(6,  "visibility",       "private",       "contextual",    "public",        "protocol"),
    Axis(7,  "mutability",       "immutable",     "unspecified",   "mutable",       "protocol"),
    Axis(8,  "directionality",   "inbound",       "undirected",    "outbound",      "protocol"),
    Axis(9,  "granularity",      "aggregate",     "unspecified",   "atomic",        "protocol"),
    Axis(10, "lifecycle",        "archived",      "active",        "draft",         "protocol"),
]

# Layer: Hardware / Platform (trits 11-20)
HARDWARE_AXES = [
    Axis(11, "compute_class",    "embedded",      "general",       "server",        "hardware"),
    Axis(12, "connectivity",     "offline",       "intermittent",  "always-on",     "hardware"),
    Axis(13, "storage_class",    "ephemeral",     "cached",        "persistent",    "hardware"),
    Axis(14, "precision",        "approximate",   "standard",      "exact",         "hardware"),
    Axis(15, "endianness",       "little",        "unspecified",   "big",           "hardware"),
    Axis(16, "word_width",       "narrow",        "standard",      "wide",          "hardware"),
    Axis(17, "security_level",   "untrusted",     "standard",      "hardened",      "hardware"),
    Axis(18, "power_state",      "sleep",         "idle",          "active",        "hardware"),
    Axis(19, "locality",         "remote",        "networked",     "local",         "hardware"),
    Axis(20, "determinism",      "stochastic",    "unspecified",   "deterministic", "hardware"),
]

# Layer: Organization / Enterprise (trits 21-35)
ORGANIZATION_AXES = [
    Axis(21, "ownership",        "external",      "shared",        "internal",      "organization"),
    Axis(22, "confidentiality",  "classified",    "standard",      "open",          "organization"),
    Axis(23, "authority",        "subordinate",   "peer",          "authoritative", "organization"),
    Axis(24, "compliance",       "non-compliant", "unassessed",    "compliant",     "organization"),
    Axis(25, "cost_center",      "expense",       "neutral",       "revenue",       "organization"),
    Axis(26, "urgency",          "deferred",      "normal",        "critical",      "organization"),
    Axis(27, "approval_state",   "rejected",      "pending",       "approved",      "organization"),
    Axis(28, "department_class", "support",       "cross-functional", "operations", "organization"),
    Axis(29, "geographic_scope", "local",         "regional",      "global",        "organization"),
    Axis(30, "retention",        "delete",        "review",        "retain",        "organization"),
    Axis(31, "audit_status",     "flagged",       "unaudited",     "cleared",       "organization"),
    Axis(32, "stakeholder",      "external",      "mixed",         "internal",      "organization"),
    Axis(33, "maturity",         "experimental",  "standard",      "proven",        "organization"),
    Axis(34, "impact",           "low",           "moderate",      "high",          "organization"),
    Axis(35, "frequency",        "rare",          "periodic",      "continuous",    "organization"),
]

# Layer: Application / Domain (trits 36-55)
APPLICATION_AXES = [
    Axis(36, "domain",           "technical",     "general",       "business",      "application"),
    Axis(37, "data_type",        "unstructured",  "semi-structured", "structured",  "application"),
    Axis(38, "flow_direction",   "outflow",       "static",        "inflow",        "application"),
    Axis(39, "value_sign",       "negative",      "zero",          "positive",      "application"),
    Axis(40, "scale",            "small",         "medium",        "large",         "application"),
    Axis(41, "relationship",     "child",         "standalone",    "parent",        "application"),
    Axis(42, "status",           "failed",        "in-progress",   "complete",      "application"),
    Axis(43, "priority",         "low",           "normal",        "high",          "application"),
    Axis(44, "source_type",      "computed",      "hybrid",        "observed",      "application"),
    Axis(45, "currency",         "stale",         "current",       "projected",     "application"),
    Axis(46, "aggregation",      "detail",        "summary",       "rollup",        "application"),
    Axis(47, "sentiment",        "negative",      "neutral",       "positive",      "application"),
    Axis(48, "action_type",      "query",         "reference",     "mutation",      "application"),
    Axis(49, "entity_class",     "event",         "abstract",      "physical",      "application"),
    Axis(50, "quantity_type",    "rate",           "absolute",      "cumulative",   "application"),
    Axis(51, "time_horizon",     "historical",    "current",       "forecast",      "application"),
    Axis(52, "validation",       "invalid",       "unvalidated",   "validated",     "application"),
    Axis(53, "version_state",    "deprecated",    "current",       "next",          "application"),
    Axis(54, "dependency",       "blocked",       "independent",   "blocking",      "application"),
    Axis(55, "format_class",     "binary",        "mixed",         "text",          "application"),
]

# Layer: Entity / Record (trits 56-70)
ENTITY_AXES = [
    Axis(56, "entity_role",      "object",        "attribute",     "subject",       "entity"),
    Axis(57, "identity_type",    "anonymous",     "pseudonymous",  "identified",    "entity"),
    Axis(58, "completeness",     "partial",       "standard",      "complete",      "entity"),
    Axis(59, "encoding",         "compressed",    "standard",      "expanded",      "entity"),
    Axis(60, "reference_type",   "symbolic",      "direct",        "semantic",      "entity"),
    Axis(61, "nullability",      "null",          "default",       "populated",     "entity"),
    Axis(62, "uniqueness",       "duplicate",     "unverified",    "unique",        "entity"),
    Axis(63, "ordering",         "unordered",     "partial",       "total",         "entity"),
    Axis(64, "indexability",     "unindexed",     "secondary",     "primary",       "entity"),
    Axis(65, "volatility",       "stable",        "moderate",      "volatile",      "entity"),
    Axis(66, "precision_level",  "approximate",   "standard",      "high-precision","entity"),
    Axis(67, "origin",           "derived",       "hybrid",        "original",      "entity"),
    Axis(68, "size_class",       "small",         "medium",        "large",         "entity"),
    Axis(69, "link_density",     "isolated",      "sparse",        "dense",         "entity"),
    Axis(70, "update_mode",      "batch",         "periodic",      "realtime",      "entity"),
]

# Layer: Field / Instance (trits 71-81)
FIELD_AXES = [
    Axis(71, "numeric_sign",     "negative",      "zero",          "positive",      "field"),
    Axis(72, "bound_type",       "lower",         "unbounded",     "upper",         "field"),
    Axis(73, "comparison",       "less_than",     "equal",         "greater_than",  "field"),
    Axis(74, "trend",            "decreasing",    "stable",        "increasing",    "field"),
    Axis(75, "anomaly",          "below_normal",  "normal",        "above_normal",  "field"),
    Axis(76, "confidence",       "low",           "medium",        "high",          "field"),
    Axis(77, "relevance",        "tangential",    "moderate",      "primary",       "field"),
    Axis(78, "recency",          "old",           "moderate",      "new",           "field"),
    Axis(79, "specificity",      "general",       "moderate",      "specific",      "field"),
    Axis(80, "affect",           "negative",      "neutral",       "positive",      "field"),
    Axis(81, "salience",         "background",    "normal",        "foreground",    "field"),
]

# Assemble all axes
ALL_AXES: List[Axis] = (
    PROTOCOL_AXES + HARDWARE_AXES + ORGANIZATION_AXES + 
    APPLICATION_AXES + ENTITY_AXES + FIELD_AXES
)

# Index by ID for quick lookup
AXIS_BY_ID: Dict[int, Axis] = {a.id: a for a in ALL_AXES}

# Group by layer
AXES_BY_LAYER: Dict[str, List[Axis]] = {}
for ax in ALL_AXES:
    AXES_BY_LAYER.setdefault(ax.layer, []).append(ax)


class SemanticDimensions:
    """Interface for working with the 81-dimensional meaning space."""
    
    def __init__(self):
        self.axes = ALL_AXES
        self.n_dims = len(self.axes)
    
    def describe(self, trits: np.ndarray) -> List[str]:
        """Human-readable description of a trit vector.
        
        Only describes non-zero dimensions (displacements from baseline).
        """
        descriptions = []
        for i, t in enumerate(trits):
            if t != 0 and i < len(self.axes):
                descriptions.append(self.axes[i].describe_trit(int(t)))
        return descriptions
    
    def layer_summary(self, trits: np.ndarray) -> Dict[str, Dict]:
        """Summarize trit vector by trust hierarchy layer."""
        summary = {}
        for layer_name, axes in AXES_BY_LAYER.items():
            active = []
            for ax in axes:
                idx = ax.id - 1  # 0-indexed
                if idx < len(trits) and trits[idx] != 0:
                    active.append(ax.describe_trit(int(trits[idx])))
            summary[layer_name] = {
                "total_dims": len(axes),
                "active_dims": len(active),
                "descriptions": active,
            }
        return summary
    
    def get_axis(self, dim: int) -> Axis:
        """Get axis definition by dimension number (1-indexed)."""
        return AXIS_BY_ID[dim]
    
    def zero_vector(self) -> np.ndarray:
        """The origin: universal baseline (all zeros)."""
        return np.zeros(self.n_dims, dtype=np.int8)

File: TSUUIDKit/Sources/TSUUIDKit/Dimensions/Dimensions.swift
import Foundation

/// A single semantic dimension axis.
public struct Axis: Sendable {
    public let id: Int
    public let name: String
    public let negative: String
    public let neutral: String
    public let positive: String
    public let layer: Layer

    public enum Layer: String, Sendable, CaseIterable {
        case protocol_ = "protocol"
        case hardware = "hardware"
        case organization = "organization"
        case application = "application"
        case entity = "entity"
        case field = "field"
    }

    public func describeTrit(_ value: Int8) -> String {
        switch value {
        case -1: return "\(name): \(negative)"
        case  0: return "\(name): \(neutral)"
        case  1: return "\(name): \(positive)"
        default: fatalError("Invalid trit value: \(value)")
        }
    }
}

/// The 81-dimensional meaning space.
/// Port of Python tsuuid.dimensions — identical axis definitions.
public enum SemanticDimensions {
    private static let protocolAxes: [Axis] = [
        Axis(id: 1,  name: "temporality",    negative: "past",       neutral: "atemporal",      positive: "future",        layer: .protocol_),
        Axis(id: 2,  name: "modality",        negative: "deletion",    neutral: "reference",      positive: "creation",      layer: .protocol_),
        Axis(id: 3,  name: "certainty",       negative: "denied",      neutral: "unknown",        positive: "confirmed",     layer: .protocol_),
        Axis(id: 4,  name: "cardinality",     negative: "singular",    neutral: "unspecified",    positive: "plural",        layer: .protocol_),
        Axis(id: 5,  name: "obligation",      negative: "forbidden",   neutral: "optional",       positive: "required",      layer: .protocol_),
        Axis(id: 6,  name: "visibility",      negative: "private",     neutral: "contextual",     positive: "public",        layer: .protocol_),
        Axis(id: 7,  name: "mutability",      negative: "immutable",   neutral: "unspecified",    positive: "mutable",       layer: .protocol_),
        Axis(id: 8,  name: "directionality",  negative: "inbound",     neutral: "undirected",     positive: "outbound",      layer: .protocol_),
        Axis(id: 9,  name: "granularity",     negative: "aggregate",   neutral: "unspecified",    positive: "atomic",        layer: .protocol_),
        Axis(id: 10, name: "lifecycle",        negative: "archived",    neutral: "active",         positive: "draft",         layer: .protocol_),
    ]

    private static let hardwareAxes: [Axis] = [
        Axis(id: 11, name: "compute_class",   negative: "embedded",    neutral: "general",        positive: "server",        layer: .hardware),
        Axis(id: 12, name: "connectivity",     negative: "offline",     neutral: "intermittent",   positive: "always-on",     layer: .hardware),
        Axis(id: 13, name: "storage_class",    negative: "ephemeral",   neutral: "cached",         positive: "persistent",    layer: .hardware),
        Axis(id: 14, name: "precision",        negative: "approximate", neutral: "standard",       positive: "exact",         layer: .hardware),
        Axis(id: 15, name: "endianness",       negative: "little",      neutral: "unspecified",    positive: "big",           layer: .hardware),
        Axis(id: 16, name: "word_width",       negative: "narrow",      neutral: "standard",       positive: "wide",          layer: .hardware),
        Axis(id: 17, name: "security_level",   negative: "untrusted",   neutral: "standard",       positive: "hardened",      layer: .hardware),
        Axis(id: 18, name: "power_state",      negative: "sleep",       neutral: "idle",           positive: "active",        layer: .hardware),
        Axis(id: 19, name: "locality",         negative: "remote",      neutral: "networked",      positive: "local",         layer: .hardware),
        Axis(id: 20, name: "determinism",      negative: "stochastic",  neutral: "unspecified",    positive: "deterministic", layer: .hardware),
    ]

    private static let organizationAxes: [Axis] = [
        Axis(id: 21, name: "ownership",        negative: "external",      neutral: "shared",           positive: "internal",      layer: .organization),
        Axis(id: 22, name: "confidentiality",  negative: "classified",    neutral: "standard",         positive: "open",          layer: .organization),
        Axis(id: 23, name: "authority",         negative: "subordinate",   neutral: "peer",             positive: "authoritative", layer: .organization),
        Axis(id: 24, name: "compliance",        negative: "non-compliant", neutral: "unassessed",       positive: "compliant",     layer: .organization),
        Axis(id: 25, name: "cost_center",       negative: "expense",       neutral: "neutral",          positive: "revenue",       layer: .organization),
        Axis(id: 26, name: "urgency",           negative: "deferred",      neutral: "normal",           positive: "critical",      layer: .organization),
        Axis(id: 27, name: "approval_state",    negative: "rejected",      neutral: "pending",          positive: "approved",      layer: .organization),
        Axis(id: 28, name: "department_class",  negative: "support",       neutral: "cross-functional", positive: "operations",    layer: .organization),
        Axis(id: 29, name: "geographic_scope",  negative: "local",         neutral: "regional",         positive: "global",        layer: .organization),
        Axis(id: 30, name: "retention",         negative: "delete",        neutral: "review",           positive: "retain",        layer: .organization),
        Axis(id: 31, name: "audit_status",      negative: "flagged",       neutral: "unaudited",        positive: "cleared",       layer: .organization),
        Axis(id: 32, name: "stakeholder",       negative: "external",      neutral: "mixed",            positive: "internal",      layer: .organization),
        Axis(id: 33, name: "maturity",          negative: "experimental",  neutral: "standard",         positive: "proven",        layer: .organization),
        Axis(id: 34, name: "impact",            negative: "low",           neutral: "moderate",         positive: "high",          layer: .organization),
        Axis(id: 35, name: "frequency",         negative: "rare",          neutral: "periodic",         positive: "continuous",    layer: .organization),
    ]

    private static let applicationAxes: [Axis] = [
        Axis(id: 36, name: "domain",           negative: "technical",     neutral: "general",          positive: "business",      layer: .application),
        Axis(id: 37, name: "data_type",        negative: "unstructured",  neutral: "semi-structured",  positive: "structured",    layer: .application),
        Axis(id: 38, name: "flow_direction",   negative: "outflow",       neutral: "static",           positive: "inflow",        layer: .application),
        Axis(id: 39, name: "value_sign",       negative: "negative",      neutral: "zero",             positive: "positive",      layer: .application),
        Axis(id: 40, name: "scale",            negative: "small",         neutral: "medium",           positive: "large",         layer: .application),
        Axis(id: 41, name: "relationship",     negative: "child",         neutral: "standalone",       positive: "parent",        layer: .application),
        Axis(id: 42, name: "status",           negative: "failed",        neutral: "in-progress",      positive: "complete",      layer: .application),
        Axis(id: 43, name: "priority",         negative: "low",           neutral: "normal",           positive: "high",          layer: .application),
        Axis(id: 44, name: "source_type",      negative: "computed",      neutral: "hybrid",           positive: "observed",      layer: .application),
        Axis(id: 45, name: "currency",         negative: "stale",         neutral: "current",          positive: "projected",     layer: .application),
        Axis(id: 46, name: "aggregation",      negative: "detail",        neutral: "summary",          positive: "rollup",        layer: .application),
        Axis(id: 47, name: "sentiment",        negative: "negative",      neutral: "neutral",          positive: "positive",      layer: .application),
        Axis(id: 48, name: "action_type",      negative: "query",         neutral: "reference",        positive: "mutation",      layer: .application),
        Axis(id: 49, name: "entity_class",     negative: "event",         neutral: "abstract",         positive: "physical",      layer: .application),
        Axis(id: 50, name: "quantity_type",    negative: "rate",          neutral: "absolute",         positive: "cumulative",    layer: .application),
        Axis(id: 51, name: "time_horizon",     negative: "historical",    neutral: "current",          positive: "forecast",      layer: .application),
        Axis(id: 52, name: "validation",       negative: "invalid",       neutral: "unvalidated",      positive: "validated",     layer: .application),
        Axis(id: 53, name: "version_state",    negative: "deprecated",    neutral: "current",          positive: "next",          layer: .application),
        Axis(id: 54, name: "dependency",       negative: "blocked",       neutral: "independent",      positive: "blocking",      layer: .application),
        Axis(id: 55, name: "format_class",     negative: "binary",        neutral: "mixed",            positive: "text",          layer: .application),
    ]

    private static let entityAxes: [Axis] = [
        Axis(id: 56, name: "entity_role",      negative: "object",        neutral: "attribute",        positive: "subject",       layer: .entity),
        Axis(id: 57, name: "identity_type",    negative: "anonymous",     neutral: "pseudonymous",     positive: "identified",    layer: .entity),
        Axis(id: 58, name: "completeness",     negative: "partial",       neutral: "standard",         positive: "complete",      layer: .entity),
        Axis(id: 59, name: "encoding",         negative: "compressed",    neutral: "standard",         positive: "expanded",      layer: .entity),
        Axis(id: 60, name: "reference_type",   negative: "symbolic",      neutral: "direct",           positive: "semantic",      layer: .entity),
        Axis(id: 61, name: "nullability",      negative: "null",          neutral: "default",          positive: "populated",     layer: .entity),
        Axis(id: 62, name: "uniqueness",       negative: "duplicate",     neutral: "unverified",       positive: "unique",        layer: .entity),
        Axis(id: 63, name: "ordering",         negative: "unordered",     neutral: "partial",          positive: "total",         layer: .entity),
        Axis(id: 64, name: "indexability",     negative: "unindexed",     neutral: "secondary",        positive: "primary",       layer: .entity),
        Axis(id: 65, name: "volatility",       negative: "stable",        neutral: "moderate",         positive: "volatile",      layer: .entity),
        Axis(id: 66, name: "precision_level",  negative: "approximate",   neutral: "standard",         positive: "high-precision",layer: .entity),
        Axis(id: 67, name: "origin",           negative: "derived",       neutral: "hybrid",           positive: "original",      layer: .entity),
        Axis(id: 68, name: "size_class",       negative: "small",         neutral: "medium",           positive: "large",         layer: .entity),
        Axis(id: 69, name: "link_density",     negative: "isolated",      neutral: "sparse",           positive: "dense",         layer: .entity),
        Axis(id: 70, name: "update_mode",      negative: "batch",         neutral: "periodic",         positive: "realtime",      layer: .entity),
    ]

    private static let fieldAxes: [Axis] = [
        Axis(id: 71, name: "numeric_sign",     negative: "negative",      neutral: "zero",             positive: "positive",      layer: .field),
        Axis(id: 72, name: "bound_type",       negative: "lower",         neutral: "unbounded",        positive: "upper",         layer: .field),
        Axis(id: 73, name: "comparison",       negative: "less_than",     neutral: "equal",            positive: "greater_than",  layer: .field),
        Axis(id: 74, name: "trend",            negative: "decreasing",    neutral: "stable",           positive: "increasing",    layer: .field),
        Axis(id: 75, name: "anomaly",          negative: "below_normal",  neutral: "normal",           positive: "above_normal",  layer: .field),
        Axis(id: 76, name: "confidence",       negative: "low",           neutral: "medium",           positive: "high",          layer: .field),
        Axis(id: 77, name: "relevance",        negative: "tangential",    neutral: "moderate",         positive: "primary",       layer: .field),
        Axis(id: 78, name: "recency",          negative: "old",           neutral: "moderate",         positive: "new",           layer: .field),
        Axis(id: 79, name: "specificity",      negative: "general",       neutral: "moderate",         positive: "specific",      layer: .field),
        Axis(id: 80, name: "affect",           negative: "negative",      neutral: "neutral",          positive: "positive",      layer: .field),
        Axis(id: 81, name: "salience",         negative: "background",    neutral: "normal",           positive: "foreground",    layer: .field),
    ]

    public static let allAxes: [Axis] =
        protocolAxes + hardwareAxes + organizationAxes +
        applicationAxes + entityAxes + fieldAxes

    public static func axis(_ id: Int) -> Axis { allAxes[id - 1] }

    public static func axes(for layer: Axis.Layer) -> [Axis] {
        allAxes.filter { $0.layer == layer }
    }

    public static func describe(_ tv: TritVector) -> [String] {
        tv.trits.enumerated().compactMap { i, t in
            guard t != 0, i < allAxes.count else { return nil }
            return allAxes[i].describeTrit(t)
        }
    }

    public static func layerSummary(_ tv: TritVector) -> [Axis.Layer: [String]] {
        var result: [Axis.Layer: [String]] = [:]
        for (i, t) in tv.trits.enumerated() where t != 0 && i < allAxes.count {
            let ax = allAxes[i]
            result[ax.layer, default: []].append(ax.describeTrit(t))
        }
        return result
    }
}
